### PR TITLE
Fix - Start both node and webpack-dev servers in dev mode for http login

### PR DESCRIPTION
**Pull request by codebyuma**
- Add concurrently and update app to start both node and webpack-dev servers at once but on different ports
- Add proxy to webpack-dev-server to redirect api calls to backend server
- Add cors to allow the /api/auth/login post request to go from port 3000 to port 8080

Connected to https://github.com/rangle/rangle-starter/issues/113

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,7 +22,7 @@ $ npm run dev
 ```
 This runs a development mode server with live reload etc. Linter warnings will be displayed with each reload.
 
-Open `http://localhost:8080` in your browser.
+Open `http://localhost:3000` in your browser.
```
Comment (SethDavenport) [inline README.md]: So in dev mode, webpack-dev-server should still run on port 8080.

The node server runs on 3000, but you'll be opening your browser to port 8080

File: package.json [hunk 2]
```diff
@@ -9,15 +9,15 @@
...
+    "node": "4.4.4"
   },
   "scripts": {
     "prebuild": "npm run clean",
     "build": "cross-env NODE_ENV=production webpack",
     "clean": "rimraf dist/",
     "cover": "remap-istanbul -i coverage/chrome/coverage-final.json -o coverage/coverage.json",
     "postcover": "istanbul report lcov text-summary",
-    "dev": "cross-env NODE_ENV=development webpack-dev-server -d --inline --progress --no-info",
+    "dev": "cross-env NODE_ENV=development concurrently 'node server/node-server.js' 'webpack-dev-server -d --inline --progress --no-info'",
```
Comment (SethDavenport) [inline package.json]: If might be worth using nodemon for debug mode, that way you get an auto-rebuild on the server side as well.

File: package.json [hunk 3]
```diff
@@ -69,8 +69,10 @@
     "basscss-typography": "^3.0.3",
     "body-parser": "^1.15.0",
     "classlist-polyfill": "^1.0.2",
+    "concurrently": "^2.0.0",
     "console-polyfill": "^0.2.2",
     "core-js": "^2.3.0",
+    "cors": "^2.7.1",
```
Comment (SethDavenport) [inline package.json]: This shouldn't be needed - that's mainly the point of the proxy config.

File: server/node-server.js [hunk 4]
```diff
@@ -61,7 +62,8 @@ passport.deserializeUser( (id, done) => {
   done(null, authPassport.getUserById(id, users));
 });
 
-app.post('/api/auth/login',
+app.options('/api/auth/login', cors({ origin: 'http://localhost:3000' }));
```
Comment (SethDavenport) [inline server/node-server.js]: Ditto

File: webpack.config.js [hunk 5]
```diff
@@ -94,8 +94,9 @@ module.exports = {
   plugins: plugins,
 
   devServer: {
+    port: 3000,
```
Comment (SethDavenport) [inline webpack.config.js]: Ok I see - you changed the dev server port instead of the node server port.

File: package.json (hunk 3 above)
Comment (codebyuma) [inline package.json]: That's what I thought but, without cors, I get this error when I try to post to the server on 8080:

`XMLHttpRequest cannot load http://localhost:8080/api/auth/login. Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:3000' is therefore not allowed access.`

**Comment by SethDavenport** [comment]
You're on the right track, however it would be nice to keep the user-facing ports the same in dev mode and prod mode.

Also CORS headers should not be needed; the proxy will make it look to the client like all requests are coming from :8080; but API ones will be proxied to 3000 under the hood.

Finally it's probably worth using `nodemon` in dev mode to have a the node server restart when server source code is changed.

So I was thinking something more like this:

package.json:
```json
scripts: {
  "dev:server": "cross-env NODE_ENV=production PORT=3000 nodemon server/node-server.js",
  "dev:client": cross-env NODE_ENV=production webpack-dev-server -d --inline --progress --no-info",
  "dev": "concurrently 'npm run dev:client' 'npm run dev:server'",
...
}
```

webpack.config.js:
```json
devServer: {
  proxy: Object.assign({}, proxy, { '*': 'http://localhost:3000" }
}
```

File: webpack.config.js (hunk 5 above)
Comment (codebyuma) [inline webpack.config.js]: Yeah. I did that so the node server runs on the same port regardless of if you're in dev or prod mode and therefore the paths in the http requests remain the same across both.

**Comment by codebyuma** [comment]
Ok, sounds good. I'll update accordingly and see if I can get this to work

**Comment by codebyuma** [comment]
The above works! I realize now I didn't quite understand how proxies work but it makes sense now.

@SethDavenport Did you intend to have the NODE_ENV=production for the dev scripts in your comment above?

**Comment by SethDavenport** [comment]
Lol no I did not - sorry

**Comment by codebyuma** [comment]
Haha, that's ok. I assumed as much and switched it to development :)

**Comment by SethDavenport** [comment]
:shipit:

**Comment by codebyuma** [comment]
Closes https://github.com/rangle/rangle-starter/issues/113

**Pull request merged**